spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/internal/spring/ArmeriaServerPortInfoApplicationContextInitializer.java

```diff
@@ -0,0 +1,121 @@
+        private static final String LOCAL_ARMERIA_PORTS = "local.armeria.ports";
+
+        private final ConfigurableApplicationContext applicationContext;
+
+        Listener(ConfigurableApplicationContext applicationContext) {
+            this.applicationContext = applicationContext;
+        }
+
+        @Override
+        public void onApplicationEvent(ArmeriaServerStartedEvent event) {
+            setPortProperty(applicationContext, event.getSource().activeLocalPort());
+            setPortProperty(applicationContext, event.getSource().activePorts().values().stream()
+                                                     .map(p -> p.localAddress().getPort())
+                                                     .collect(toList()));
+        }
+
+        private void setPortProperty(ApplicationContext context, int port) {
+            if (context instanceof ConfigurableApplicationContext) {
+                setPortProperty(((ConfigurableApplicationContext) context).getEnvironment(), port);
+            }
+            if (context.getParent() != null) {
+                setPortProperty(context.getParent(), port);
+            }
+        }
+
+        private void setPortProperty(ConfigurableEnvironment environment, int port) {
+            final MutablePropertySources sources = environment.getPropertySources();
+            PropertySource<?> source = sources.get("server.ports");
+            if (source == null) {
+                source = new MapPropertySource("server.ports", new HashMap<>());
+                sources.addFirst(source);
+            }
+            setPortProperty(port, source);
+        }
+
+        @SuppressWarnings("unchecked")
+        private void setPortProperty(int port, PropertySource<?> source) {
+            ((Map<String, Object>) source.getSource()).put(LOCAL_ARMERIA_PORT, port);
+        }
+
+        private void setPortProperty(ApplicationContext context, List<Integer> ports) {
+            if (context instanceof ConfigurableApplicationContext) {
+                setPortProperty(((ConfigurableApplicationContext) context).getEnvironment(), ports);
+            }
+            if (context.getParent() != null) {
+                setPortProperty(context.getParent(), ports);
+            }
+        }
+
+        private void setPortProperty(ConfigurableEnvironment environment, List<Integer> ports) {
+            final MutablePropertySources sources = environment.getPropertySources();
+            PropertySource<?> source = sources.get("server.ports");
+            if (source == null) {
+                source = new MapPropertySource("server.ports", new HashMap<>());
+                sources.addFirst(source);
+            }
+            setPortProperty(ports, source);
+        }
+
+        @SuppressWarnings("unchecked")
+        private void setPortProperty(List<Integer> ports, PropertySource<?> source) {
+            ((Map<String, Object>) source.getSource()).put(LOCAL_ARMERIA_PORTS, ports);
+        }
```

Comment:
Fixed. Do you mean adding a static keyword?

Reply:
* This method is used in only one place.
* There are many ways of method overloading for `setPortProperty` in this file.

So I thought it would be nice to embed this code.😀